### ISSUE TITLE
Query failure on master branch

### ISSUE THREAD

**Issue by Eulerizeit** [greenplum-db/gpdb#3446]
The following query works in `5X_STABLE` but fails on master branch.

```SQL
SELECT 1
FROM (
  SELECT a, count(*) OVER (PARTITION BY a) FROM (VALUES (1,1)) AS foo(a)
) AS sup(c, d)
WHERE c = 87 ;
```

The error on the screen:
```
FATAL:  Unexpected internal error (planwindow.c:1730)
DETAIL:  FailedAssertion("!(cdbpathlocus_is_valid(*_locus))", File: "planwindow.c", Line: 1730)
HINT:  Process 24500 will wait for gp_debug_linger=120 seconds before termination.
Note that its locks and other resources will not be released until then.
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
The connection to the server was lost. Attempting reset: Succeeded.
```

**Comment by hsyuan** [comment]
Are they the same failure? https://github.com/greenplum-db/gpdb/issues/3243

**Comment by hlinnaka** [comment]
Yeah, it's the same failure. Reproducable on GPDB 5 as well.

**Comment by Eulerizeit** [comment]
@d @oarap it this exists on Master and on 5.x and MADlib is working on 5.x we clearly need to get this green then spend a bit more time trying to repro the error they are getting.

**Comment by jchampio** [comment]
A [piece of `ensure_collocation_and_order`](https://github.com/greenplum-db/gpdb/blob/master/src/backend/optimizer/plan/planwindow.c#L1708-L1731) doesn't handle the case where we have an empty/`NIL` set of distribution path keys. @orap and I are trying a naive fix (just don't call `CdbPathLocus_MakeHashed` in that situation), and we'll see if that breaks anything.

**Comment by jchampio** [comment]
@hlinnaka We're pretty sure our naive fix (which passes ICG and fixes the assertion) is wrong. It seems like we need to assign the `output_locus` if `cdbpathlocus_collocates` returns false, but we don't have a `dist_pathkeys` list during this query that we can construct a new locus around. Any thoughts on how we can move forward?

**Comment by hlinnaka** [comment]
So, the problem is that `assure_collocation_and_order` is confused on whether there's partitioning going on, or not. There is a PARTITION BY clause, so it falls into the "yes, let's hash distribute" code path rather than the "no, let's gather" code path. But then dist_pathkeys is empty, so it fails to actually create a valid hash locus.

I think we need to construct dist_pathkeys earlier, and make the "is there partitioning going on" based on whether dist_pathkeys is empty or not. Opened PR for that: https://github.com/greenplum-db/gpdb/pull/3471

**Comment by hlinnaka** [comment]
Fixed by PR #3471.
